### PR TITLE
[SofaPython][PSDE] ADD: Switching to cpp-style datatypes in the PSDE factory

### PR DESCRIPTION
**Pull request by marques-bruno**
Part 1 of #629 This PR creates more Sofa Types in the PSDEFactory and uses a cpp-style typename convention 

______________________________________________________

This PR: 
- [x] builds with SUCCESS for all platforms on the CI.
- [x] does not generate new warnings.
- [x] does not generate new unit test failures.
- [x] does not generate new scene test failures.
- [x] does not break API compatibility.
- [x] is more than 1 week old (or has fast-merge label).

**Reviewers will merge only if all these checks are true.**

### CODE REVIEW

**Comment by damienmarchal** [comment]
Hi Bruno,

Thanks a lot for this PR that is really needed and made in a nice way. I will have a look if it complex to add a deprecation message in the binding so that when the user is using the "old" convention something is printed and replacement string is suggested.

**Review by jnbrunet — COMMENTED** [review]
Seems good to me. Maybe just update the addNewData method's documentation at the end of the binding file to take into account those new type aliases. Also, adding a way to easily get the list of accepted types from python would also be useful. Although I'm not sure of the best way to achieve this...

**Comment by marques-bruno** [comment]
@jnbrunet a list of accepted types would be cool indeed. Maybe by adding a method+binding in the factory class that returns a list of all its keys in its registry? Though maybe more appropriate in a separate PR.
I'll update the doc accordingly though!

**Comment by damienmarchal** [comment]
It may be overkill on this specific PR but when string matching are involved I like to use difflib (https://docs.python.org/2/library/difflib.html) to compute  approximated match. This allows me to propose suggestion to users when they are misspelling something eg:
"vec3" will suggest "Vec3d"

(Actually we should probably use it in the binding part when querying for a Data<> or getChild/getObject)

**Comment by marques-bruno** [comment]
@damienmarchal I think it's a great idea, and it seems to fit well to this PR, now that we have more that 4 scalar types in the factory...
Did you use it already somewhere in sofa?

**Comment by damienmarchal** [comment]
I used it in PSL but it was on the python side:
 matches = difflib.get_close_matches("MyStrzng", ["MyStrong", "MyString", "ABCD", "EFGH" ], n=4)

To call it from C I assume that some CPython ugliness work has to be done.

**Comment by marques-bruno** [comment]
or https://github.com/duckie/difflib or equivalent, if the licence allows..?

File: applications/plugins/SofaPython/Binding_Base.cpp [hunk 1]
```diff
@@ -47,20 +52,131 @@ typedef sofa::helper::Factory< std::string, BaseData> PSDEDataFactory;
...
+                    "Mat4x4d", new DataCreator<sofa::defaulttype::Mat4x4d>());
+        s_localfactory->registerCreator(
+                    "Mat2x2f", new DataCreator<sofa::defaulttype::Mat2x2f>());
+        s_localfactory->registerCreator(
+                    "Mat3x3f", new DataCreator<sofa::defaulttype::Mat3x3f>());
+        s_localfactory->registerCreator(
+                    "Mat4x4f", new DataCreator<sofa::defaulttype::Mat4x4f>());
+
+        // Topology
+        s_localfactory->registerCreator("Edge", new DataCreator<Tetra>());
```
Comment (tgaugry) [inline applications/plugins/SofaPython/Binding_Base.cpp]: Why not `Edge`, `Triangle` and so on ?

Comment (damienmarchal) [inline applications/plugins/SofaPython/Binding_Base.cpp]: Oups.

File: applications/plugins/SofaPython/Binding_Base.cpp [hunk 2]
```diff
@@ -47,20 +52,131 @@ typedef sofa::helper::Factory< std::string, BaseData> PSDEDataFactory;
...
+        s_localfactory->registerCreator("Triangle", new DataCreator<Tetra>());
```
Comment (damienmarchal) [inline applications/plugins/SofaPython/Binding_Base.cpp]: Oups I did it again.

File: applications/plugins/SofaPython/Binding_Base.cpp [hunk 3]
```diff
@@ -47,20 +52,131 @@ typedef sofa::helper::Factory< std::string, BaseData> PSDEDataFactory;
...
+        s_localfactory->registerCreator("Quad", new DataCreator<Tetra>());
```
Comment (damienmarchal) [inline applications/plugins/SofaPython/Binding_Base.cpp]: :)

**Comment by guparan** [comment]
[ci-build][with-scene-tests]

**Pull request merged**